hub/core/chunk_engine.py

```diff
@@ -558,28 +564,67 @@ def _sanitize_samples(self, samples):
             samples = list(samples)
         return samples, verified_samples
 
+    def can_fit_to_next(self, samples, fit_row):
+        chunk_id = self.chunk_id_encoder.get_next_chunk_id(fit_row)
+        if chunk_id is None:
+            return False
+
+        chunk_name = ChunkIdEncoder.name_from_id(chunk_id)  # type: ignore
+        chunk_commit_id = self.get_chunk_commit(chunk_name)
+        chunk_key = get_chunk_key(self.key, chunk_name, chunk_commit_id)
+        chunk_size = self.cache.get_object_size(chunk_key)
+
+        sample_bytes = 0
+        for _, sample in enumerate(samples):
+            sample_bytes += len(sample.buffer)
+
+        sum_size = chunk_size + sample_bytes
+        if sum_size < self.min_chunk_size:
+            return True
+        return False
+
     def _samples_to_chunks(
         self,
         samples,
         start_chunk: Optional[BaseChunk] = None,
         register: bool = True,
         update_commit_diff: bool = False,
+        append_to_end: bool = True,
+        extend: bool = True,
+        fit_row: Optional[int] = None,
     ):
+        # TODO check if there is a suboptimal chunk that we can use
+        # if not then create new next to current one and set data
         current_chunk = start_chunk
+
         updated_chunks = []
         if current_chunk is None:
-            current_chunk = self._create_new_chunk(register)
+            current_chunk = self._create_new_chunk(register, row=fit_row)
             updated_chunks.append(current_chunk)
         enc = self.chunk_id_encoder
         tiles = {}
         nsamples = len(samples)
         if register and update_commit_diff:
             commit_diff = self.commit_diff
         while len(samples) > 0:
-            num_samples_added = current_chunk.extend_if_has_space(samples)  # type: ignore
+            num_samples_added = current_chunk.extend_if_has_space(
+                samples, extend=extend, end=append_to_end
+            )  # type: ignore
             self.register_new_creds(num_samples_added, samples)
             if num_samples_added == 0:
-                current_chunk = self._create_new_chunk(register)
+                if fit_row is not None and self.can_fit_to_next(samples, fit_row):
+                    chunk_id = self.chunk_id_encoder.get_next_chunk_id(fit_row)
+                    next_chunk = self.get_chunk_from_chunk_id(int(chunk_id))  # type: ignore
+                    return self._samples_to_chunks(
+                        samples,
+                        start_chunk=next_chunk,
+                        register=True,
+                        update_commit_diff=True,
+                        append_to_end=False,
+                        extend=False,
+                        fit_row=fit_row + 1,
```

Comment:
why +1?

Reply:
this is `row+1` as we are trying to insert on the new chunk, I will change it to be new variable